Add smart logic to log information about plugin status changes

gsoldevila opened the pull request:
## Summary

New attempt at fixing https://github.com/elastic/kibana/issues/116718
Inspired on https://github.com/elastic/kibana/pull/126320

Here's what the newly logged `[status]` information looks like on a fresh startup:
<img width="1834" alt="image" src="https://github.com/elastic/kibana/assets/25349407/d78d7f88-139f-4daf-9dc0-c4e6724ea412">

The first 2 entries are logs from Core services 🆕 .
The next 5 entries are emitted due to `taskManager` plugin emitting a degraded status right at startup.
I have created an issue to tackle that one: https://github.com/elastic/kibana/issues/168237

elasticmachine commented:
Pinging @elastic/kibana-core (Team:Core)

gsoldevila commented on x-pack/plugins/licensing/server/plugin_status.ts:
```diff
@@ -5,17 +5,16 @@
   stop$: Observable<void>
 ): Observable<ServiceStatus> => {
   return license$.pipe(
-    startWith(undefined),
```
No need to start emitting meaningless statuses from this plugin, `plugins_status.ts` already has a default `'unavailable'` status.

gsoldevila commented on packages/core/status/core-status-server-internal/src/status_service.ts:
```diff
@@ -184,6 +185,12 @@ export class StatusService implements CoreService<InternalStatusServiceSetup> {
+      getPluginStatusChangesMessages(statusChanges).forEach((message) => {
+        this.logger.info(message);
+      });
+    });
```
This is the entry point to the PR. We start observing plugins statuses and logging meaningful information whenever there are changes.

jloleysens replied:
Very cool, overall look good so far!

---

What do you think about combining this into a single, multiline log?

Or, if we're worried about breaking some monitoring regexes that are looking for a status change message maybe keep the single message and then have the "digest" message second (so max two messages)?

My concern is potential interleaving of messages.

Just a thought, need to still do rest of review

gsoldevila replied:
Good point!

IMO on one hand, some of the messages can be logged independently; they just appear together because the _Observables_ we observe are debouncing events internally.

On the other hand, there are others, like details of each plugin status change, which do belong together.
I have updated these in https://github.com/elastic/kibana/pull/168207/commits/320933d74efa7187168549671f0ebc268547edbf to create some multiline logs.

gsoldevila commented on packages/core/status/core-status-server-internal/src/log_plugins_status.test.ts:
```diff
@@ -0,0 +1,585 @@
+          unavailable: [],
+        },
+      })
+    ).toMatchInlineSnapshot(`
```
In this test we have a very explanatory example of how the "diff status" events are processed, and what things are logged.

gsoldevila commented on packages/core/status/core-status-server-internal/src/log_plugins_status.ts:
```diff
@@ -0,0 +1,185 @@
+  stop$: Observable<void>
+): Observable<PluginStatusChanges> => {
+  return plugins$.pipe(
+    startWith({}), // consider all plugins unavailable by default
```
Otherwise we are silently ignoring single emission, which could be meaningful.

gsoldevila commented on packages/core/status/core-status-server-internal/src/log_plugins_status.ts:
```diff
@@ -0,0 +1,185 @@
+    takeUntil(stop$),
+    pairwise(),
+    map(([oldStatus, newStatus]) => getPluginsStatusDiff(oldStatus, newStatus)),
+    filter((statusChanges) => statusChanges.updated > 0)
```
We only emit events if at least one plugin status has changed.
NB we no longer _debounce_ or _throttleTime_ here,
as we rely on https://github.com/elastic/kibana/pull/128324 improvements to have a lot less emissions.

pgayvallet replied:
> NB we no longer debounce or throttleTime here

It's great to know we don't need it anymore performance-wise, but I still wonder if keeping debounce/throttling with a small time range wouldn't be better to avoid the edge case of a "blinking" plugin flooding the logs, wdyt?

gsoldevila replied:
I am not sure `debounce` or `throttleTime` are the best candidates for what we want to achieve:
- `debounce` only emits the most recent event when the time comes, and the others are simply lost.
- `throttleTime` also ignores events for some time after a given emission.

I was thinking of a combination of `map` + `filter` using an external temporary cache to keep track of number of emissions of each _pluginName_, and filter them up to a certain time or number of repetitions, having some sort of factorised message:`Plugin A switched status 38 times the last minute:

jloleysens reviewed (commented):
Left a comment, don't need to block on it

gsoldevila commented:
I triggered a `kibana-apis-capacity-testing` build as requested:
https://buildkite.com/elastic/kibana-apis-capacity-testing/builds/2147

jloleysens commented on packages/core/status/core-status-server-internal/src/log_plugins_status.test.ts:
```diff
@@ -0,0 +1,583 @@
+
+          unavailable: [],
+        },
+        total: 2,
+        updated: 2,
```
I know this is a test fixture, but are these numbers correct here? Or am I misunderstanding what they mean

gsoldevila replied:
Good catch! I wrote them in the beginning and then forgot to update them when I added more entries. Will update.

jloleysens reviewed (commented):
Tested locally and seems to be working as intended!

Left a couple of non-blocker comments.

---

I see we now have logs, as you indicated, like:

```
[2023-10-10T15:05:32.390+02:00][INFO ][status] The following plugins are now available: alerting, fleet, licensing (and 48 more)
 - 'alerting' is now available: Alerting is (probably) ready
 - 'fleet' is now available: Fleet is setting up
 - 'licensing' is now available: License fetched
 
```

Does the `and 48 more` mean there are 48 other plugins that are now available? Would we similarly truncate these "reason" summaries if these 48 plugins went unavailable?

---

Would be great to get @afharo 's input too since he originally proposed the enhancement I believe!

gsoldevila commented on packages/core/status/core-status-common/src/service_status.ts:
```diff
@@ -37,6 +37,11 @@ export interface ServiceStatus<Meta extends Record<string, any> | unknown = unkn
+  /**
+   * Whether this status is reported by the service/plugin itself, or it is inferred from the status of other services that this service depends on
+   */
+  isReportedStatus?: boolean;
```
From the status point of view, there are 2 types of plugins:
* Those that register a `status$` Observable and proactively report their status (there are 4 of them RN).
* Those that do NOT register a `status$` Observable, and their status is inferred from the status of their dependencies (plugin dependency tree) + status of core services: (elasticsearch client, saved objects).

This flag is basically telling us whether the status is directly reported or it is inferred.

From a logging perspective, the `(and 48 more)` refers to 48 other plugins that have changed status because of their dependencies, but they don't have a "self" status themselves (thus, not much value logging them).

Hope that answers your [other question](https://github.com/elastic/kibana/pull/168207#pullrequestreview-1667544494) too.

gsoldevila replied:
Thinking about it, plugin owners should not need to worry about this property, it's managed internally by us, will check if I can extend the interface with an internal one.

pgayvallet commented on packages/core/status/core-status-server-internal/src/log_plugins_status.ts:
```diff
@@ -0,0 +1,187 @@
+    const previousLevel = previous[pluginName]?.level ?? ServiceStatusLevels.unavailable;
+    if (previousLevel !== pluginStatus.level) {
+      // this plugin status has changed
+      updates[currentLevel.toString()].push({
+        pluginName,
+        ...pluginStatus,
+      } as PluginStatusWithName);
```
NIT: we're constructing the exact same spreaded structure twice (`{ pluginName, ...pluginStatus }`). It can be reused between line 6 and 65 instead.

pgayvallet commented on packages/core/status/core-status-server-internal/src/log_plugins_status.ts:
```diff
@@ -0,0 +1,187 @@
+          messages.push(`${statusReportingPlugins[0].pluginName} is ${currentLevel}: ${reason}`);
+        } else {
+          messages.push(
+            `'${statusReportingPlugins[0].pluginName}' plugin (and some others that depend on it) are ${currentLevel}: ${reason}`
```
NIT: can we be more precise than `some` here? (and does it make sense?)

pgayvallet commented on packages/core/status/core-status-server-internal/src/log_plugins_status.ts:
```diff
@@ -0,0 +1,187 @@
+      } else if (statusReportingPlugins.length > 1) {
+        if (inferredCount === 0) {
+          messages.push(
+            `The following plugins are ${currentLevel}: ${statusReportingPlugins.map(
```
NIT: message inconsistency between
- `statusReportingPlugins.length === 1` => `{pluginName} is ${currentLevel}`
- `statusReportingPlugins.length > 1` => `The following plugins are ${currentLevel}`.

I think more consistency would make sense, e.g `The following plugin...` and `The following plugins...`

gsoldevila replied:
When there is a single plugin changing to a specific status, I am putting the `${reason}` in the same line, thus the distinction.

I will try to unify and preserve consistency.

pgayvallet replied:
It's a NIT, honestly, but I'm mostly thinking about us here, and I think it would help us for log discovery (e.g searching for "the following" or things like that.

afharo replied:
Mixing this comment and the `some` ones... from o11y POV, we want to know which plugin(s) caused the unavailability. From a user perspective, I would expect something like:
* `Kibana is now unavailable (was available). [#] plugins are now unavailable. Status change caused by plugin(s) ${listOfPluginsThatCausedTheDegradedState}.`

As an extra, I find it useful that we are printing the reason for the plugins turning unavailable (maybe it can be printed via the logger `status.${plugin}`?) Although, based on recent experience, we may risk logging information that shouldn't be logged 🙈, so I don't know anymore 🤷.

If it's OK to log the reason, I wonder if the status could be placed as part of the meta of the "general" log so, one log entry would have: 
1. The message: as defined above: summary of the global status, the number of plugins unavailable + which plugins caused the status change
2. The meta: it contains the status at that point in time

WDYT?

gsoldevila replied:
Sounds like a desirable log, although there are a few challenges to achieve this.
Kibana status change can be caused by degraded plugins but also by degraded core services (saved objects, elasticsearch client). We will need a way to find out which of them caused the status change.

pgayvallet commented on packages/core/status/core-status-server-internal/src/log_plugins_status.ts:
```diff
@@ -0,0 +1,187 @@
+          );
+        } else {
+          messages.push(
+            `The following plugins (and some others that depend on them) are ${currentLevel}: ${statusReportingPlugins.map(
```
NIT: `some` again, can we display the actual number instead, and would that be useful?

pgayvallet replied:
> The problem is that I don't have information of what plugins depend on the one that is e.g. degraded.
I only know they are also degraded, but they might depend on other plugins that just switched to degraded.

Yeah but we're displaying a single message for all the degraded plugins (`The following pluginS`), right? so even if we don't know which one(s) caused the "child" plugins to degrade, we still know that it was all caused by one, or more non-inferred plugins?

gsoldevila replied:
There's a distinction between:
1. _The following plugins are NOW `degraded`_ ... in this scenario I know that the "child" ones have just become degraded, caused by the non-inferred ones.
2. _The following plugins are `degraded`_ ... in this scenario, I know that some plugins are degraded, but I don't know if they were already degraded or they have just become degraded. I could perhaps do a `diff` and remove the newly degraded ones, which will be logged in the other message (1). Or I could simply not log the non-healthy ones that haven't changed status.

pgayvallet replied:
Okay, then we may want to take some sync time to talk about it, as I thought we were only displaying messages for the first case (when plugin statuses effectively change)

pgayvallet commented on packages/core/status/core-status-server-internal/src/log_plugins_status.ts:
```diff
@@ -0,0 +1,187 @@
+        }
+
+        statusReportingPlugins.forEach(({ pluginName, ...pluginStatus }) =>
+          messages.push(`${pluginName} is ${currentLevel}: ${getReason(pluginStatus)}`)
```
Related to the `The following plugins...` vs ``${pluginName} is ...` discussion

pgayvallet reviewed (commented):
Overall looking good. Left a few comments and questions

afharo reviewed (commented):
Overall LGTM. Not approving because there are a few comments from other reviewers. 

Also, I'd like to discuss the possibility of improving the summary log entry 😇

gsoldevila commented:
I updated the logic + tests and I think I cover all remarks now:
* Having an overall status that tells what services and/or plugins are causing that status.
* Logging messages ONLY for plugins that **change** status since last emission.
* Not logging figures (e.g. _and 152 more_) for inferred plugins' statuses.
* Added logging for core services individual status changes.
* Used appropriate contextualized loggers (`status.core` vs `status.plugins`)

See screenshot below, highlighting the status messages logged at startup:
<img width="1834" alt="image" src="https://github.com/elastic/kibana/assets/25349407/d78d7f88-139f-4daf-9dc0-c4e6724ea412">

pgayvallet reviewed (approved):
Thanks for supporting us 3 all giving different opinions and nit picking every possible details.

I really like the final outcome, LGTM

gsoldevila commented on x-pack/plugins/task_manager/server/plugin.ts:
```diff
@@ -179,7 +179,7 @@ export class TaskManagerPlugin
       combineLatest([core.status.derivedStatus$, serviceStatus$]).pipe(
         map(([derivedStatus, serviceStatus]) =>
-          serviceStatus.level > derivedStatus.level ? serviceStatus : derivedStatus
+          serviceStatus.level >= derivedStatus.level ? serviceStatus : derivedStatus
```
This way, if `taskManager` status is the same level as core services, the taskManager message will be reported.

gsoldevila commented:
Launched an APIs capacity testing with the latest changes
https://buildkite.com/elastic/kibana-apis-capacity-testing/builds/2268

afharo reviewed (approved):
LGTM! Thanks for iterating! 🧡
